src/parse.rs

```diff
@@ -37,18 +41,174 @@ impl Expr {
+                // evaluate the `with` statement earlier than needed. Instead
+                // we create ExprSource::Ident expressions then put those in
+                // the attribute set.
+                for inherit in set.inherits() {
+                    // Handle syntax like `inherit (some_expression) foo` by
+                    // rewriting it to `foo = some_expression.foo`, allowing
+                    // `some_expression` to be lazily evaluated.
+                    if let Some(from) = inherit.from() {
+                        let from = Gc::new(Self::parse(
+                            from.inner().ok_or(ERR_PARSING)?,
+                            new_scope.clone(),
+                        )?);
+
+                        // For our example described above, add `some_expression`,
+                        // `foo`, and `bar` to the ExprSource so they're all visible
+                        // to interactive tooling.
+                        inherits.push(Ok(from.clone()));
+
+                        for ident in inherit.idents() {
+                            let name = ident.as_str();
+                            let index = Box::new(Expr {
+                                value: GcCell::new(None),
+                                source: ExprSource::Ident {
+                                    name: name.to_string(),
+                                },
+                                range: None,
+                                scope: scope.clone(),
+                            });
+                            let attr = Gc::new(Expr {
+                                value: GcCell::new(None),
+                                source: ExprSource::Select {
+                                    from: Ok(from.clone()),
+                                    index: Ok(index),
+                                },
+                                range: Some(ident.node().text_range()),
+                                scope: scope.clone(),
+                            });
+                            inherits.push(Ok(attr.clone()));
+                            map.insert(name.to_string(), attr);
+                        }
+                    } else {
+                        // Handle `inherit` from scope
+                        for ident in inherit.idents() {
+                            let name = ident.as_str();
+                            let attr = Gc::new(Expr {
+                                value: GcCell::new(None),
+                                source: ExprSource::Ident {
+                                    name: name.to_string(),
+                                },
+                                range: Some(ident.node().text_range()),
+                                scope: scope.clone(),
+                            });
+                            inherits.push(Ok(attr.clone()));
+                            map.insert(name.to_string(), attr);
```

Comment:
This doesn't seem to deal with the case `{ inherit a; a = 2; }` which normally results in an evaluation error. In that case we shouldn't provide any value, otherwise a user may expect something that isn't correct.

Reply:
Agreed. I've added an error message so the evaluator now shows "attribute `a` defined more than once" for that example code